KAFKA-17774: Adding capability to handle max fetch records in Share Fetch (KIP-932)

apoorvmittal10 opened the pull request:
The PR adds capability to restrict the messages in Share Fetch. The max fetch records will be an additional way to limit the number of records sent from broker to client.

In Share Fetch, with min and mx bytes, there exists 3 problems:
1. The max.poll.records client config sends the max number of records defined to application but might have fetched extra becuase of higher max bytes. But the timeout for the sent records has started on the broker.
2. As the application processes records as per max.poll.records, hence those number of records are sent in every acknowledgement. This causes the cache data to be tracked per offset as the batch is broken.
3. The client has to sent the partial acknoledgment batch and cannot piggyback on fetch requests.

To handle the above scenario max fetch records has been added. Once this PR is merged and we define the right methodolgy then KIP will be updated to have max fetch records in share fetch RPC rather as broker config.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-17774: Max fetch records for share fetch (KIP-932)

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 committed:
Max fetch records implementation and tests

apoorvmittal10 committed:
Remove temp changes

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
         long firstOffset,
-        long lastOffset
+        long lastOffset,
+        int maxMessages
```
We've been using `maxFetchRecords` so far, so I don't think we should be using `messages` now. `maxRecords` would be better.

apoorvmittal10 replied:
Done, changed to maxFetchRecords. I missed earlier while renaming.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
+
+            // Check how many messages can be acquired from the batch.
+            long lastAcquiredOffset = lastOffset;
+            if (maxMessages < (lastAcquiredOffset - firstAcquiredOffset + 1)) {
```
nit: I think you can lose the inner parentheses for readability.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroupConfig.java:
```diff
@@ -79,7 +85,8 @@ public class ShareGroupConfig {
             .define(SHARE_GROUP_PARTITION_MAX_RECORD_LOCKS_CONFIG, INT, SHARE_GROUP_PARTITION_MAX_RECORD_LOCKS_DEFAULT, between(100, 10000), MEDIUM, SHARE_GROUP_PARTITION_MAX_RECORD_LOCKS_DOC)
-            .define(SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_CONFIG, INT, SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_DEFAULT, MEDIUM, SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_DOC);;
+            .define(SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_CONFIG, INT, SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_DEFAULT, MEDIUM, SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_DOC)
+            .define(SHARE_FETCH_MAX_FETCH_RECORDS_CONFIG, INT, SHARE_FETCH_MAX_FETCH_RECORDS_DEFAULT, HIGH, SHARE_FETCH_MAX_FETCH_RECORDS_DOC);
```
`defineInternal`

apoorvmittal10 replied:
Thanks, done.

AndrewJSchofield commented on share/src/main/java/org/apache/kafka/server/share/fetch/ShareAcquiredRecords.java:
```diff
@@ -0,0 +1,73 @@
+        return count;
+    }
+
+    public static ShareAcquiredRecords emptyShareAcquiredRecords() {
```
I'd just call it `empty()`.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on share/src/main/java/org/apache/kafka/server/share/fetch/ShareAcquiredRecords.java:
```diff
@@ -0,0 +1,73 @@
+
+    public static ShareAcquiredRecords fromAcquiredRecords(AcquiredRecords acquiredRecords) {
+        return new ShareAcquiredRecords(
+            Collections.singletonList(acquiredRecords),
```
`List.of`

apoorvmittal10 replied:
Yeah Java 8 is gone, done.

AndrewJSchofield reviewed (changes requested):
This looks like a good first step towards experimenting with improving the batching. A few minor comments, but overall looks good to me.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
@AndrewJSchofield Thanks for the review. I have address the comments.

apoorvmittal10 committed:
Removing not required change

apoorvmittal10 committed:
Fixing config test

adixitconfluent commented on core/src/test/java/kafka/server/share/DelayedShareFetchTest.java:
```diff
@@ -229,12 +233,12 @@ public void testReplicaManagerFetchShouldHappenOnComplete() {
                 new FetchParams(ApiKeys.SHARE_FETCH.latestVersion(), FetchRequest.ORDINARY_CONSUMER_ID, -1, MAX_WAIT_MS,
                         1, 1024 * 1024, FetchIsolation.HIGH_WATERMARK, Optional.empty()), groupId, Uuid.randomUuid().toString(),
-                new CompletableFuture<>(), partitionMaxBytes);
+                new CompletableFuture<>(), partitionMaxBytes, 100);
```
Can we use this 100 as a constant, I see its usage at many places in this file and other test files.

apoorvmittal10 replied:
Done.

adixitconfluent commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -433,6 +433,7 @@ class BrokerServer(
         config.shareGroupConfig.shareGroupDeliveryCountLimit,
         config.shareGroupConfig.shareGroupPartitionMaxRecordLocks,
         config.shareGroupConfig.shareFetchPurgatoryPurgeIntervalRequests,
+        config.shareGroupConfig.shareFetchMaxFetchRecords(),
```
nit: Just being consistent with the above values, we can get rid of the parenthesis.

apoorvmittal10 replied:
Thanks for pointing, my bad. Done.

adixitconfluent commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1209,34 +1352,42 @@ public void testAcquireReleasedRecordMultipleBatches() {
-        acquiredRecordsList = sharePartition.acquire(
+        acquiredRecordsList = fetchAcquiredRecords(sharePartition.acquire(
             MEMBER_ID,
+            Integer.MAX_VALUE,
```
Same comment as the 100 one, can we put Integer.MAX_VALUE as a variable for DEFAULT_MAX_FETCH_RECORDS

apoorvmittal10 replied:
Done.

adixitconfluent reviewed (commented):
Overall PR feels fine to me. Left a few minor comments.

junrao commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -219,6 +227,7 @@ private SharePartitionManager(
         this.delayedActionsQueue = delayedActionsQueue;
         this.groupConfigManager = groupConfigManager;
         this.shareGroupMetrics = new ShareGroupMetrics(Objects.requireNonNull(metrics), time);
+        this.maxFetchRecords = shareFetchMaxFetchRecords;
```
Instead of duplicating the code between two constructors. Could we just forward the constructor to the other constructor?

apoorvmittal10 replied:
You are right, it's just that delayedPurgatory uses the same timer with lock currently hence when I moved to `this` then it becomes hard. I ll do this refactor when PR to move purgatory to replica manger is merged. This is on my list.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
         lock.writeLock().lock();
         try {
+            // If same batch is fetched and previous batch is removed from the cache then we need to
+            // update the batch first offset to endOffset, only if enfOffset is passed the firstOffset.
```
typo enfOffset
enfOffset is passed => enfOffset passed

apoorvmittal10 replied:
My bad, done.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
+                // limit the acquired records. The last offset shall be the batches last offset
+                // which falls under the max messages limit. As the max fetch records is the soft
+                // limit, the last offset can be higher than the max messages.
+                lastAcquiredOffset = findLastOffsetFromBatchWithRequestOffset(batches, firstAcquiredOffset + maxFetchRecords - 1);
```
`firstAcquiredOffset + maxFetchRecords - 1` won't be accurate for compacted topics. Not sure if there is a simple solution though.

apoorvmittal10 replied:
Yeah for compacted it might be not accurate. We have one open issue to handle and test for compacted topics. Compacted topics complete support might be 4.1 release.

junrao commented on share/src/main/java/org/apache/kafka/server/share/fetch/ShareAcquiredRecords.java:
```diff
@@ -0,0 +1,72 @@
+        this.count = 0;
+    }
+
+    public List<AcquiredRecords> records() {
```
records => acquiredRecords ?

apoorvmittal10 replied:
I initially kept as `acquiredRecords` but the moved to `records` as the class name is ShareAcquiredRecords. Though moved it back to `acquiredRecords`.

junrao commented on core/src/main/java/kafka/server/share/ShareFetchUtils.java:
```diff
@@ -84,24 +88,31 @@ static Map<TopicIdPartition, ShareFetchResponseData.PartitionData> processFetchR
                 } else {
                     partitionData
+                        // We set the records to the fetchPartitionData records. We do not alter the records
+                        // fetched from the replica manager as they follow zero copy buffer. The acquired records
```
If `records` contains 10 batches and we only acquire records in the first batch, it's wasteful to return all 10 batches in the records.

apoorvmittal10 replied:
Yeah, that's right and hence I wrote this comment. We ll be experimenting with max bytes and max fetch records to see how less wasteful the response can be with configurations.

chia7712 replied:
Out of curiosity, how should users configure max.poll.records to minimize network I/O waste? Should users set `max.poll.records=Infinite`? If so, should we consider making an infinite value the default for `max.poll.records`?

apoorvmittal10 replied:
`max.poll.records=Infinite` has it's own downsides as the locks on server might timeout prior to client completed processing of all records, if the records returned are more than the time capacity in which they can be processed by client. Hence there has to be a balance with `fetch max bytes` and `max.poll.records`. We are testing with what seems to be the best behaviour in which application can define these configurations easier.

AndrewJSchofield replied:
@chia7712 The aim is to drive this from the consumer. For some workloads, it will be critical not to over-fetch records because the processing time is significant and we just risk lots of acquisition lock timeouts and wasted fetched records. For others, we want pre-fetching to allow overlapping of fetching and processing. This PR lets us start to experiment with this area. I'll make a partner PR for the consumer soon (but not yet changing the protocol because that will be the final step when we have decided how to do it nicely).

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
+            // to 0 but firstOffset can be higher than 0.
+            long firstAcquiredOffset = firstOffset;
+            if (cachedState.isEmpty() && endOffset > firstAcquiredOffset) {
+                firstAcquiredOffset = endOffset;
```
It's possible for `endOffset > firstAcquiredOffset` even when `cachedState.isEmpty()` is false, right? In that case, we still want to bound firstAcquiredOffset by endOffset.

apoorvmittal10 replied:
The situation where we need to move the firstAcquiredOffset is only when we receive first ever shar partition acquire request. In that case, cache shall be empty. Also we should only move the offset if endOffset is passed the firstOffset.

junrao replied:
Is that true? Suppose endOffset is initialized to 105 and the log has two batches with offset range [100, 110] and [111, 120]. Initially cachedState is empty and we call acquireNewBatchRecords with the first batch. We will set firstAcquiredOffset to 105 in this case. [105, 110] will be in acquired state and endOffset advances to 111. Another share fetch request comes in and we call acquireNewBatchRecords with the second batch. [111, 120] will be in acquired state and endOffset advances to 121. The first batch times out. [105, 110] will be in available state and endOffset is reset to 105. Now another share fetch request comes in. We will call acquireNewBatchRecords with the first batch. Since cachedState is not empty, we won't set firstAcquiredOffset to 105 and leave it as 100, which seems incorrect.

apoorvmittal10 replied:
> Since cachedState is not empty, we won't set firstAcquiredOffset to 105 and leave it as 100, which seems incorrect.

Yes, that's fine. As in those cases we need not to move firstAcquired offset as we are not handling fresh batch acquisition i.e. this check in only important when we are acquiring the batch for first time.

This method will not be triggered as submap is not empty, i.e. replica manager batch [100, 100] will have an overlap with [105, 110] cached batch.

The complete batch from [105,110] will be acquired as per the lines from 633, where inflight batch is a full match and available to acquire.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Made a pass of non-testing files. Added a few comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 commented:
@adixitconfluent @junrao @AndrewJSchofield Can I please get the review again. I am rebasing the PR with upstream merge.

adixitconfluent reviewed (approved):
Thanks for the PR @apoorvmittal10 . LGTM.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1059,36 +1085,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
         lock.writeLock().lock();
         try {
+            // If same batch is fetched and previous batch is removed from the cache then we need to
+            // update the batch first offset to endOffset, only if endOffset is passed the firstOffset.
```
endOffset is passed => endOffset passed

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. A couple of more comments.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 committed:
Correcting code comment

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -513,18 +514,27 @@ public long nextFetchOffset() {
     ) {
         log.trace("Received acquire request for share partition: {}-{} memberId: {}", groupId, topicIdPartition, memberId);
+        if (maxFetchRecords <= 0) {
+            // Nothing to acquire.
```
I guess this is temporary, right? we should not allow `maxFetchRecords<=0` in the future.

apoorvmittal10 replied:
It's a safe check in the method. As there could be multiple partitions in a single share fetch request and a single partition might satisfy the max fetch records criteria hence skip acquiring for other.

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1658,6 +1706,26 @@ private long findLastOffsetAcknowledged() {
+        long offset
+    ) {
+        for (RecordBatch batch : batches) {
+            if (offset >= batch.baseOffset() && offset <= batch.lastOffset()) {
```
`lastOffset` loads the entire batch header, so perhaps we should avoid checking `lastOffset` for every batch." For example:
```java
        RecordBatch previous = null;
        for (RecordBatch batch : batches) {
            if (offset >= batch.baseOffset()) {
                previous = batch;
                continue;
            }
            break;
        }
        if (previous != null && offset <= previous.lastOffset()) return previous.lastOffset();
        return offset;
```

apoorvmittal10 replied:
That's a great suggestion.

Just for my understanding, I was wondering if really `lastOffset` call does additional loading i.e looking at implementation of DefaultRecordBatch it seems lastOffset is just getting additional information from already loaded buffer. I might be missing something, hence can you please help me learn more so I can optimize code at other places as well.

chia7712 replied:
> Just for my understanding, I was wondering if really lastOffset call does additional loading i.e looking at implementation of DefaultRecordBatch it seems lastOffset is just getting additional information from already loaded buffer. I might be missing something, hence can you please help me learn more so I can optimize code at other places as well.

If I recall correctly, the records from `ReplicaManager#readFromLog` are `FileRecords`. Therefore, they don't pre-load the header or entire batch until we call a method like lastOffset.

https://github.com/apache/kafka/blob/d36eb9ed3f0ba3d9b58671f7bd74f82b7bbf6b9e/clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java#L685

junrao replied:
We iterate the batches from `FileRecords` when serving share fetch requests. The iteration logic in `FileLogInputStream` already loads the header.

```
    public FileChannelRecordBatch nextBatch() throws IOException {
        FileChannel channel = fileRecords.channel();
        if (position >= end - MIN_HEADER_READ_REQUIRED)
            return null;

        final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(MAX_HEADER_READ_ATTEMPTED);
        Utils.readFully(channel, logHeaderBuffer, position);
        logHeaderBuffer.flip();

```

chia7712 replied:
> We iterate the batches from FileRecords when serving share fetch requests. The iteration logic in FileLogInputStream already loads the header.

It seems that only the `magic` and `offset` (`logHeaderBuffer`) are pre-loaded, while lastOffset is excluded.

apoorvmittal10 replied:
I debugged further to check that `DefaultFileChannelRecordBatch` is created which loads the header when lastOffset is called hence implemented the proposal from @chia7712. Please let us know if that's not right understanding @junrao.

chia7712 replied:
I open https://issues.apache.org/jira/browse/KAFKA-18989 to optimize `FileRecord#searchForOffsetWithSize` too

AndrewJSchofield replied:
@apoorvmittal10 ^^

apoorvmittal10 replied:
Thanks, this was on my list and was discussing same with @junrao, couple of weeks back. I shall review the PR, thanks @chia7712 for surfacing it.

apoorvmittal10 replied:
@chia7712 so with this ticket we are optimizing the lastBatch usage in `FileRecord#searchForOffsetWithSize` but not optimizing if lastOffset of the batch can somehow be pre-loaded with loading headers?

chia7712 replied:
> but not optimizing if lastOffset of the batch can somehow be pre-loaded with loading headers?

Given that lastOffsetDelta is located at byte 23, it necessitates seeking the file channel to read specific bytes. Furthermore, we may iterate through numerous batches if the index is sparse. I'm uncertain whether incurring this cost is justified. WDYT?

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -612,16 +634,20 @@ public List<AcquiredRecords> acquire(
                     .setFirstOffset(inFlightBatch.firstOffset())
                     .setLastOffset(inFlightBatch.lastOffset())
                     .setDeliveryCount((short) inFlightBatch.batchDeliveryCount()));
+                acquiredCount += (int) (inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1);
```
I suppose this `acquiredCount may also be inaccurate if the topic is compacted, right?

apoorvmittal10 replied:
Yeah, but server can't determine that as we do not read the records content in server. Rather clients inform for any gaps in the record batch.

chia7712 reviewed (commented):
@apoorvmittal10 thanks for this fix.

> To handle the above scenario max fetch records has been added. Once this PR is merged and we define the right methodolgy then KIP will be updated to have max fetch records in share fetch RPC rather as broker config.

`max.poll.records` behaves differently in a shared consumer, making it seem like a separate configuration. Users will likely need to reconsider the value they used for normal consumers, right? If so, have we considered encouraging users to focus on max bytes instead and ignoring this config in a shared consumer?

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 commented:
@AndrewJSchofield @chia7712 @junrao Thanks for reviewing, appreciate efforts on the long PR. I have rebased the PR and addressed the comments, can I please get the review again.

junrao commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroupConfig.java:
```diff
@@ -71,15 +72,21 @@ public class ShareGroupConfig {
     public static final String SHARE_FETCH_PURGATORY_PURGE_INTERVAL_REQUESTS_DOC = "The purge interval (in number of requests) of the share fetch request purgatory";
 
+    // Broker temporary configuration to limit the number of records fetched by a share fetch request.
+    public static final String SHARE_FETCH_MAX_FETCH_RECORDS_CONFIG = "share.fetch.max.fetch.records";
```
Not sure if this should be tied the the max.poll.records on the client side. It seems that our goal is to bound the number of un-acked records per consumer instance, which may not always match max.poll.records on the client side. For example, if the client polls two max.poll.records without acking, it doesn't mean that we should allow the client to have 2 * max.poll.records un-acked records. Perhaps we could add a new group level config like `group.max.record.locks.per.member`?

apoorvmittal10 replied:
That's a good feedback @junrao. We might not neccessarily tie exactly as per max poll records but if we do have this functionality in broker then can figure out what best works for us with different configurations defined on the client.

AndrewJSchofield replied:
@junrao I'll add your suggestion into the mix. I anticipate some of this work will extend into 4.1.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Left one more comment.

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the reply. The PR LGTM. @chia7712 and @AndrewJSchofield: Any more comments from you?

apoorvmittal10 commented:
@AndrewJSchofield @chia7712 Please let me know if need to address anything in this PR or it's good to merge.

chia7712 commented on share/src/main/java/org/apache/kafka/server/share/fetch/ShareAcquiredRecords.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class ShareAcquiredRecords {
+
+    public static final ShareAcquiredRecords EMPTY_SHARE_ACQUIRED_RECORDS = new ShareAcquiredRecords();
```
it is used internally, so could you please change it from public to private?

apoorvmittal10 replied:
Done.

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1138,6 +1185,7 @@ private void acquireSubsetBatchRecords(
         } finally {
             lock.writeLock().unlock();
         }
+        return result.size();
```
Apologies, but why don't we return the count of AcquiredRecords added to the result in this method? Since result is reused by the caller, result.size won't equal the count of the subset, correct?

apoorvmittal10 replied:
Sorry, I did not understand this comment. The idea is to keep `acquireSubsetBatchRecords` method to return count of acquired records. Currently in this method we add result for every offset hence the count is the size of the result, but going forward we have to address the TODO in the method which says to batch to contiguous offsets. Once done then result size will not be same as count, then from this methid we ll specifically return the count of the acquired records.

chia7712 replied:
> Currently in this method we add result for every offset hence the count is the size of the result

We're using the same results object when calling `acquireSubsetBatchRecords`, so results.size will return the total count of all added offsets across each call. Does the condition `if (!fullMatch || inFlightBatch.offsetState() != null)` get matched only once while looping through `subMap`?
```
                    int acquiredSubsetCount = acquireSubsetBatchRecords(memberId, firstBatch.baseOffset(), lastBatch.lastOffset(), inFlightBatch, result);
                    acquiredCount += acquiredSubsetCount;
```

apoorvmittal10 replied:
You are so right, my bad. I have fixed this. While working further on this I will change some existing subset tests to account for these cases as well. Thanks for catching, I have pushed the change.

apoorvmittal10 replied:
@chia7712 Please let me know if we there is any other feedback to address, or we are good to go?

chia7712 reviewed (commented):
@apoorvmittal10 thanks for updates. Two small questions remains. PTAL

apoorvmittal10 committed:
Removing last offset iteration

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into max-fetch-records

apoorvmittal10 commented:
@chia7712 I have addressed the comments. Thanks for the review.

apoorvmittal10 committed:
Correcting subset acquired count

AndrewJSchofield reviewed (approved):
Thanks for the PR. lgtm.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1658,6 +1708,33 @@ private long findLastOffsetAcknowledged() {
+     * @param offset The request offset to find.
+     * @return The last offset of the batch which contains the request offset, otherwise the request offset.
+     */
+    private long findLastOffsetFromBatchWithRequestOffset(
```
findLastOffsetFromBatchWithRequestOffset => lastOffsetFromBatchWithRequestOffset?

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Just a minor comment.

apoorvmittal10 committed:
Renaming method

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the updated PR. LGTM
